--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -892,6 +_,12 @@ public abstract class AbstractContainerMenu {
         } else {
             float f = 0.0F;
 
+            // DivineMC start - lithium: sleeping_block_entity
+            if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity && container instanceof net.caffeinemc.mods.lithium.api.inventory.LithiumInventory inventory) {
+                return net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(inventory).getSignalStrength(container);
+            }
+            // DivineMC end - lithium: sleeping_block_entity
+
             for (int i = 0; i < container.getContainerSize(); i++) {
                 ItemStack item = container.getItem(i);
                 if (!item.isEmpty()) {