--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlockState.java
@@ -1,5 +_,6 @@
 package org.bukkit.craftbukkit.block;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.base.Preconditions;
 import java.lang.ref.WeakReference;
 import java.util.List;
@@ -14,6 +_,7 @@
 import org.bukkit.block.Block;
 import org.bukkit.block.BlockState;
 import org.bukkit.block.data.BlockData;
+import org.bukkit.craftbukkit.CraftChunk;
 import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
 import org.bukkit.craftbukkit.util.CraftLocation;
@@ -124,7 +_,7 @@
     @Override
     public Chunk getChunk() {
         this.requirePlaced();
-        return this.world.getChunkAt(this.getX() >> 4, this.getZ() >> 4);
+        return new CraftChunk(this.world.getHandle(), this.getX() >> 4, this.getZ() >> 4); // ShreddedPaper - do not sync-load from BlockState#getChunk
     }
 
     public void setData(net.minecraft.world.level.block.state.BlockState data) {
@@ -216,6 +_,7 @@
         if (!this.isPlaced()) {
             return true;
         }
+        TickThread.ensureTickThread(this.world.getHandle(), this.position, "BlockState update");
         LevelAccessor access = this.getWorldHandle();
         CraftBlock block = this.getBlock();
 
