--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -325,11 +_,14 @@
                     _chatExecutorMaxSize = _chatExecutorCoreSize;
                 }
 
-                java.util.concurrent.ThreadPoolExecutor executor = (java.util.concurrent.ThreadPoolExecutor) net.minecraft.server.MinecraftServer.getServer().chatExecutor;
-                executor.setCorePoolSize(_chatExecutorCoreSize);
-                executor.setMaximumPoolSize(_chatExecutorMaxSize);
-            }
+                // DivineMC start - Virtual Threads
+                if (net.minecraft.server.MinecraftServer.getServer().chatExecutor instanceof java.util.concurrent.ThreadPoolExecutor executor) {
+                    executor.setCorePoolSize(_chatExecutorCoreSize);
+                    executor.setMaximumPoolSize(_chatExecutorMaxSize);
+                }
+                // DivineMC end - Virtual Threads
+            }
         }
         public int maxJoinsPerTick = 5;
         public boolean sendFullPosForItemEntities = false;
         public boolean loadPermissionsYmlBeforePlugins = true;
