--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -66,6 +_,7 @@ public class Tadpole extends AbstractFish {
     );
     public boolean ageLocked; // Paper
     private org.purpurmc.purpur.controller.WaterMoveControllerWASD purpurController; // Purpur - Ridables
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Tadpole(EntityType<? extends AbstractFish> type, Level level) {
         super(type, level);
@@ -138,13 +_,16 @@ public class Tadpole extends AbstractFish {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("tadpoleBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("tadpoleActivityUpdate");
-        TadpoleAi.updateActivity(this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("tadpoleActivityUpdate");
+            TadpoleAi.updateActivity(this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     public static AttributeSupplier.Builder createAttributes() {