--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +_,8 @@
      */
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
+    Key BRAND_SHREDDEDPAPER_ID = Key.key("multipaper", "shreddedpaper"); // ShreddedPaper legacy
+    Key BRAND_ASTATINE_ID = Key.key("astatine", "astatine"); // Astatine
     /**
      * Gets the {@code ServerBuildInfo}.
      *
