--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -113,6 +_,7 @@ public class Axolotl extends Animal implements Bucketable {
     public final BinaryAnimator onGroundAnimator = new BinaryAnimator(10, EasingType.IN_OUT_SINE);
     public final BinaryAnimator movingAnimator = new BinaryAnimator(10, EasingType.IN_OUT_SINE);
     private static final int REGEN_BUFF_BASE_DURATION = 100;
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Axolotl(EntityType<? extends Axolotl> type, Level level) {
         super(type, level);
@@ -374,16 +_,19 @@ public class Axolotl extends Animal implements Bucketable {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("axolotlBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("axolotlActivityUpdate");
-        AxolotlAi.updateActivity(this);
-        profilerFiller.pop();
-        if (!this.isNoAi()) {
-            Optional<Integer> memory = this.getBrain().getMemory(MemoryModuleType.PLAY_DEAD_TICKS);
-            this.setPlayingDead(memory.isPresent() && memory.get() > 0);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("axolotlActivityUpdate");
+            AxolotlAi.updateActivity(this);
+            profilerFiller.pop();
+            if (!this.isNoAi()) {
+                Optional<Integer> memory = this.getBrain().getMemory(MemoryModuleType.PLAY_DEAD_TICKS);
+                this.setPlayingDead(memory.isPresent() && memory.get() > 0);
+            }
         }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     public static AttributeSupplier.Builder createAttributes() {