--- a/net/minecraft/world/entity/ai/sensing/HoglinSpecificSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/HoglinSpecificSensor.java
@@ -5,6 +_,7 @@
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BlockTags;
@@ -14,6 +_,7 @@
 import net.minecraft.world.entity.ai.memory.NearestVisibleLivingEntities;
 import net.minecraft.world.entity.monster.hoglin.Hoglin;
 import net.minecraft.world.entity.monster.piglin.Piglin;
+import net.minecraft.world.level.block.state.BlockState;
 
 public class HoglinSpecificSensor extends Sensor<Hoglin> {
     @Override
@@ -60,6 +_,10 @@
     }
 
     private Optional<BlockPos> findNearestRepellent(ServerLevel level, Hoglin hoglin) {
-        return BlockPos.findClosestMatch(hoglin.blockPosition(), 8, 4, pos -> level.getBlockState(pos).is(BlockTags.HOGLIN_REPELLENTS));
+        return BlockPos.findClosestMatch(hoglin.blockPosition(), 8, 4, pos -> {
+            if (!TickThread.isTickThreadFor(level, pos)) return false; // ShreddedPaper - avoid off-owner repellent scans
+            BlockState blockState = level.getBlockStateIfLoaded(pos); // ShreddedPaper - avoid sync chunk loads during repellent scans
+            return blockState != null && blockState.is(BlockTags.HOGLIN_REPELLENTS);
+        });
     }
 }
