--- a/src/main/java/com/destroystokyo/paper/network/PaperLegacyStatusClient.java
+++ b/src/main/java/com/destroystokyo/paper/network/PaperLegacyStatusClient.java
@@ -1,6 +_,7 @@
 package com.destroystokyo.paper.network;

 import com.destroystokyo.paper.event.server.PaperServerListPingEvent;
+import io.multipaper.shreddedpaper.privacy.AnonymousMode;
 import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
 import net.minecraft.ChatFormatting;
 import net.minecraft.server.MinecraftServer;
@@ -24,7 +_,7 @@

     @Override
     public InetSocketAddress getAddress() {
-        return this.address;
+        return AnonymousMode.redactPublicSocketAddress(this.address);
     }

     @Override
