--- a/src/main/java/com/destroystokyo/paper/network/PaperNetworkClient.java
+++ b/src/main/java/com/destroystokyo/paper/network/PaperNetworkClient.java
@@ -2,2 +2,3 @@
 
+import io.multipaper.shreddedpaper.privacy.AnonymousMode;
 import java.net.InetSocketAddress;
@@ -17,3 +18,3 @@
     public InetSocketAddress getAddress() {
-        return (InetSocketAddress) this.networkManager.getRemoteAddress();
+        return AnonymousMode.redactPublicSocketAddress((InetSocketAddress) this.networkManager.getRemoteAddress());
     }
