--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -114,6 +_,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     private float spinningAnimationTicks0;
     public boolean forceDancing = false; // CraftBukkit
     private org.purpurmc.purpur.controller.FlyingMoveControllerWASD purpurController; // Purpur - Ridables
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Allay(EntityType<? extends Allay> type, Level level) {
         super(type, level);
@@ -277,13 +_,16 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("allayBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("allayActivityUpdate");
-        AllayAi.updateActivity(this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("allayActivityUpdate");
+            AllayAi.updateActivity(this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override