--- a/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
@@ -4,2 +4,4 @@
 import java.util.Optional;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
@@ -32,3 +34,6 @@
     protected boolean canStillUse(ServerLevel level, Villager entity, long gameTime) {
-        return entity.getBrain().hasMemoryValue(MemoryModuleType.POTENTIAL_JOB_SITE);
+        return entity.getBrain()
+            .getMemory(MemoryModuleType.POTENTIAL_JOB_SITE)
+            .filter(globalPos -> globalPos.dimension() == level.dimension())
+            .isPresent();
     }
@@ -47,2 +52,3 @@
             if (level1 != null) {
+                ShreddedPaper.ensureSync(level1, blockPos, () -> { // ShreddedPaper - run on poi's region
                 PoiManager poiManager = level1.getPoiManager();
@@ -52,3 +58,4 @@
 
-                level.debugSynchronizers().updatePoi(blockPos);
+                level1.debugSynchronizers().updatePoi(blockPos);
+                }); // ShreddedPaper - run on poi's region
             }
