--- a/net/minecraft/world/entity/ai/behavior/JumpOnBed.java
+++ b/net/minecraft/world/entity/ai/behavior/JumpOnBed.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
+import net.minecraft.world.level.block.state.BlockState;
 import org.jspecify.annotations.Nullable;
 
 public class JumpOnBed extends Behavior<Mob> {
@@ -101,7 +_,8 @@
     }
 
     private boolean isBed(ServerLevel level, BlockPos pos) {
-        return level.getBlockState(pos).is(BlockTags.BEDS);
+        BlockState blockState = level.getBlockStateIfLoaded(pos); // ShreddedPaper - avoid sync chunk loads from baby villager bed checks
+        return blockState != null && blockState.is(BlockTags.BEDS);
     }
 
     private Optional<BlockPos> getNearestBed(Mob mob) {
