--- a/net/minecraft/network/PacketProcessor.java
+++ b/net/minecraft/network/PacketProcessor.java
@@ -88,6 +_,8 @@
             if (this.listener.shouldHandleMessage(this.packet)) {
                 try {
                     this.packet.handle(this.listener);
+                } catch (net.minecraft.server.RunningOnDifferentThreadException ignored) {
+                    return;
                 } catch (Exception var3) {
                     if (var3 instanceof ReportedException reportedException && reportedException.getCause() instanceof OutOfMemoryError) {
                         throw PacketUtils.makeReportedException(var3, this.packet, this.listener);
