--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -223,6 +_,9 @@
                 );
             }
 
+            if (this.shreddedpaper$shouldDeferProjectileMove(this.getDeltaMovement())) {
+                return;
+            }
             hitResult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
         }
 
