--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -96,6 +_,7 @@ public class Goat extends Animal {
     private static final boolean DEFAULT_HAS_RIGHT_HORN = true;
     private boolean isLoweringHead;
     private int lowerHeadTick;
+    private int behaviorTick = 0; // DivineMC - Dynamic Activation of Brain
 
     public Goat(EntityType<? extends Goat> type, Level level) {
         super(type, level);
@@ -229,13 +_,16 @@ public class Goat extends Animal {
     protected void customServerAiStep(ServerLevel level) {
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("goatBrain");
-        if (getRider() == null || !this.isControllable()) // Purpur - only use brain if no rider
-        this.getBrain().tick(level, this);
-        profilerFiller.pop();
-        profilerFiller.push("goatActivityUpdate");
-        GoatAi.updateActivity(this);
-        profilerFiller.pop();
-        super.customServerAiStep(level);
+        // DivineMC start - Dynamic Activation of Brain
+        if ((getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) { // Purpur - only use brain if no rider
+            this.getBrain().tick(level, this);
+            profilerFiller.pop();
+            profilerFiller.push("goatActivityUpdate");
+            GoatAi.updateActivity(this);
+            profilerFiller.pop();
+            super.customServerAiStep(level);
+        }
+        // DivineMC end - Dynamic Activation of Brain
     }
 
     @Override