--- a/net/minecraft/world/level/chunk/PalettedContainer.java
+++ b/net/minecraft/world/level/chunk/PalettedContainer.java
@@ -264,6 +_,28 @@ public class PalettedContainer<T> implements PaletteResize<T>, PalettedContainer
             data.palette.read(buffer, this.strategy.globalMap());
             buffer.readFixedSizeLongArray(data.storage.getRaw());
             this.data = data;
+            // DivineMC start - ModernFix: compact_bit_storage
+            if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.useCompactBitStorage && _byte > 1) {
+                long[] storageArray = this.data.storage.getRaw();
+                boolean empty = true;
+                for (long l : storageArray) {
+                    if (l != 0) {
+                        empty = false;
+                        break;
+                    }
+                }
+                if (empty && storageArray.length > 0) {
+                    T value;
+                    try {
+                        value = this.data.palette.valueFor(0);
+                    } catch (RuntimeException e) {
+                        return;
+                    }
+                    this.data = this.createOrReuseData(null, 0);
+                    this.data.palette.idFor(value, this);
+                }
+            }
+            // DivineMC end - ModernFix: compact_bit_storage
             this.addPresetValues(); // Paper - Anti-Xray - Add preset values (inefficient, but this isn't used by the server)
             this.updateData(this.data); // Paper - optimise palette reads
         } finally {