--- a/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/AmphibiousPathNavigation.java
@@ -3,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.pathfinder.AmphibiousNodeEvaluator;
 import net.minecraft.world.level.pathfinder.PathFinder;
 import net.minecraft.world.phys.Vec3;
@@ -12,9 +_,25 @@
         super(mob, level);
     }
 
+    // DivineMC start - Async Pathfinding
+    private static final org.bxteam.divinemc.async.pathfinding.NodeEvaluatorGenerator nodeEvaluatorGenerator = (org.bxteam.divinemc.async.pathfinding.NodeEvaluatorFeatures nodeEvaluatorFeatures) -> {
+        AmphibiousNodeEvaluator nodeEvaluator = new AmphibiousNodeEvaluator(false);
+        nodeEvaluator.setCanPassDoors(nodeEvaluatorFeatures.canPassDoors());
+        nodeEvaluator.setCanFloat(nodeEvaluatorFeatures.canFloat());
+        nodeEvaluator.setCanWalkOverFences(nodeEvaluatorFeatures.canWalkOverFences());
+        nodeEvaluator.setCanOpenDoors(nodeEvaluatorFeatures.canOpenDoors());
+        return nodeEvaluator;
+    };
+    // DivineMC end - Async Pathfinding
+
     @Override
     protected PathFinder createPathFinder(int maxVisitedNodes) {
         this.nodeEvaluator = new AmphibiousNodeEvaluator(false);
+        // DivineMC start - Async Pathfinding
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.asyncPathfinding) {
+            return new PathFinder(this.nodeEvaluator, maxVisitedNodes, nodeEvaluatorGenerator);
+        }
+        // DivineMC end - Async Pathfinding
         return new PathFinder(this.nodeEvaluator, maxVisitedNodes);
     }
 
@@ -40,7 +_,8 @@
 
     @Override
     public boolean isStableDestination(BlockPos pos) {
-        return !this.level.getBlockState(pos.below()).isAir();
+        BlockState blockState = this.level.getBlockStateIfLoaded(pos.below()); // ShreddedPaper - avoid sync chunk loads from amphibious navigation stability checks
+        return blockState != null && !blockState.isAir();
     }
 
     @Override
