--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -499,12 +_,17 @@ public class Commands {
     }
 
     // Fixed pool, but with discard policy
     public static final java.util.concurrent.ExecutorService COMMAND_SENDING_POOL = new java.util.concurrent.ThreadPoolExecutor(
         2, 2, 0, java.util.concurrent.TimeUnit.MILLISECONDS,
         new java.util.concurrent.LinkedBlockingQueue<>(),
-        new com.google.common.util.concurrent.ThreadFactoryBuilder()
-            .setNameFormat("Paper Async Command Builder Thread Pool - %1$d")
-            .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER))
-            .build(),
+        org.bxteam.divinemc.config.DivineConfig.VirtualThreadsCategory.useCommandBuilderScheduler()
+            ? Thread.ofVirtual()
+                .name("Paper Async Command Builder Thread Pool - ", 1)
+                .uncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER))
+                .factory()
+            : new com.google.common.util.concurrent.ThreadFactoryBuilder()
+                .setNameFormat("Paper Async Command Builder Thread Pool - %1$d")
+                .setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER))
+                .build(),
         new java.util.concurrent.ThreadPoolExecutor.DiscardPolicy()
-    );
+    ); // DivineMC - Virtual Threads
