--- a/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
@@ -22,7 +_,7 @@ import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import org.slf4j.Logger;
 
-public class ChiseledBookShelfBlockEntity extends BlockEntity implements ListBackedContainer {
+public class ChiseledBookShelfBlockEntity extends BlockEntity implements ListBackedContainer, net.caffeinemc.mods.lithium.api.inventory.LithiumTransferConditionInventory { // DivineMC - lithium: sleeping_block_entity
     public static final int MAX_BOOKS_IN_STORAGE = 6;
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int DEFAULT_LAST_INTERACTED_SLOT = -1;
@@ -171,4 +_,11 @@ public class ChiseledBookShelfBlockEntity extends BlockEntity implements ListBac
     public void removeComponentsFromTag(ValueOutput output) {
         output.discard("Items");
     }
+
+    // DivineMC start - lithium: sleeping_block_entity
+    @Override
+    public boolean lithium$itemInsertionTestRequiresStackSize1() {
+        return true;
+    }
+    // DivineMC end - lithium: sleeping_block_entity
 }