--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -28,7 +_,13 @@ import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import org.jspecify.annotations.Nullable;
 
-public class HopperBlockEntity extends RandomizableContainerBlockEntity implements Hopper {
+// DivineMC start - lithium: sleeping_block_entity
+import net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper;
+import net.caffeinemc.mods.lithium.common.hopper.LithiumStackList;
+import net.caffeinemc.mods.lithium.api.inventory.LithiumInventory;
+// DivineMC end - lithium: sleeping_block_entity
+
+public class HopperBlockEntity extends RandomizableContainerBlockEntity implements Hopper, net.caffeinemc.mods.lithium.common.block.entity.SleepingBlockEntity, net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementListener, net.caffeinemc.mods.lithium.api.inventory.LithiumInventory, net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeListener, net.caffeinemc.mods.lithium.common.hopper.UpdateReceiver, net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker { // DivineMC - lithium: sleeping_block_entity
     public static final int MOVE_ITEM_SPEED = 8;
     public static final int HOPPER_CONTAINER_SIZE = 5;
     private static final int[][] CACHED_SLOTS = new int[54][];
@@ -119,6 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
     @Override
     public void setBlockState(BlockState blockState) {
+        if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity && this.level != null && !this.level.isClientSide() && blockState.getValue(HopperBlock.FACING) != this.getBlockState().getValue(HopperBlock.FACING)) this.invalidateCachedData(); // DivineMC - lithium: sleeping_block_entity
         super.setBlockState(blockState);
         this.facing = blockState.getValue(HopperBlock.FACING);
     }
@@ -137,6 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             boolean result = tryMoveItems(level, pos, state, blockEntity, () -> {
                 return suckInItems(level, blockEntity);
             });
+            if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity) blockEntity.checkSleepingConditions(); // DivineMC - lithium: sleeping_block_entity
             if (!result && blockEntity.level.spigotConfig.hopperCheck > 1) {
                 blockEntity.setCooldown(blockEntity.level.spigotConfig.hopperCheck);
             }
@@ -199,6 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 if (flag) {
                     blockEntity.setCooldown(level.spigotConfig.hopperTransfer); // Spigot
                     setChanged(level, pos, state);
+                    if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity && !blockEntity.isOnCooldown() && !blockEntity.isSleeping() && !state.getValue(HopperBlock.ENABLED)) blockEntity.lithium$startSleeping(); // DivineMC - lithium: sleeping_block_entity
                     return true;
                 }
             }
@@ -219,11 +_,11 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
     // Paper start - Perf: Optimize Hoppers
     public static boolean skipHopperEvents;
-    private static boolean skipPullModeEventFire;
-    private static boolean skipPushModeEventFire;
+    private static final ThreadLocal<Boolean> skipPullModeEventFireThreadLocal = ThreadLocal.withInitial(() -> false); // ShreddedPaper - use thread local for skipPullModeEventFire
+    private static final ThreadLocal<Boolean> skipPushModeEventFireThreadLocal = ThreadLocal.withInitial(() -> false); // ShreddedPaper - use thread local for skipPushModeEventFire
 
     private static boolean hopperPush(final Level level, final Container destination, final Direction direction, final HopperBlockEntity hopper) {
-        skipPushModeEventFire = skipHopperEvents;
+        skipPushModeEventFireThreadLocal.set(skipHopperEvents); // ShreddedPaper - use thread local for skipPushModeEventFire
         boolean foundItem = false;
         for (int i = 0; i < hopper.getContainerSize(); ++i) {
             final ItemStack item = hopper.getItem(i);
@@ -238,7 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
                 // We only need to fire the event once to give protection plugins a chance to cancel this event
                 // Because nothing uses getItem, every event call should end up the same result.
-                if (!skipPushModeEventFire) {
+                if (!skipPushModeEventFireThreadLocal.get()) { // ShreddedPaper - use thread local for skipPushModeEventFire
                     movedItem = callPushMoveEvent(destination, movedItem, hopper);
                     if (movedItem == null) { // cancelled
                         origItemStack.setCount(originalItemCount);
@@ -274,7 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         container.setChanged(); // original logic always marks source inv as changed even if no move happens.
         movedItem.setCount(movedItemCount);
 
-        if (!skipPullModeEventFire) {
+        if (!skipPullModeEventFireThreadLocal.get()) { // ShreddedPaper - use thread local for skipPullModeEventFire
             movedItem = callPullMoveEvent(hopper, container, movedItem);
             if (movedItem == null) { // cancelled
                 origItemStack.setCount(originalItemCount);
@@ -294,9 +_,9 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
             }
 
-            ignoreBlockEntityUpdates = true;
+            ignoreBlockEntityUpdatesThreadLocal.set(true); // ShreddedPaper - use thread local for ignoreBlockEntityUpdates
             container.setItem(i, origItemStack);
-            ignoreBlockEntityUpdates = false;
+            ignoreBlockEntityUpdatesThreadLocal.remove(); // ShreddedPaper - use thread local for ignoreBlockEntityUpdates
             container.setChanged();
             return true;
         }
@@ -320,7 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         );
         final boolean result = event.callEvent();
         if (!event.calledGetItem && !event.calledSetItem) {
-            skipPushModeEventFire = true;
+            skipPushModeEventFireThreadLocal.set(true); // ShreddedPaper - use thread local for skipPushModeEventFire
         }
         if (!result) {
             applyCooldown(hopper);
@@ -343,7 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         final io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent event = new io.papermc.paper.event.inventory.PaperInventoryMoveItemEvent(sourceInventory, org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), destination, false);
         final boolean result = event.callEvent();
         if (!event.calledGetItem && !event.calledSetItem) {
-            skipPullModeEventFire = true;
+            skipPullModeEventFireThreadLocal.set(true); // ShreddedPaper - use thread local for skipPullModeEventFire
         }
         if (!result) {
             applyCooldown(hopper);
@@ -375,6 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private static void applyCooldown(final Hopper hopper) {
         if (hopper instanceof HopperBlockEntity blockEntity && blockEntity.getLevel() != null) {
             blockEntity.setCooldown(blockEntity.getLevel().spigotConfig.hopperTransfer);
+            blockEntity.skipNextSleepCheckAfterCooldown = true; // DivineMC - lithium: sleeping_block_entity
         }
     }
 
@@ -418,12 +_,25 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     // Paper end - Perf: Optimize Hoppers
 
     private static boolean ejectItems(Level level, BlockPos pos, HopperBlockEntity blockEntity) {
-        Container attachedContainer = getAttachedContainer(level, pos, blockEntity);
+        Container attachedContainer = org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity ? blockEntity.getInsertInventory(level) : getAttachedContainer(level, pos, blockEntity); // DivineMC - lithium: sleeping_block_entity
         if (attachedContainer == null) {
             return false;
         } else {
             Direction opposite = blockEntity.facing.getOpposite();
+            // DivineMC start - lithium: sleeping_block_entity
+            if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity) {
+                Boolean res = lithiumInsert(level, pos, blockEntity, attachedContainer);
+                if (res != null) {
+                    return res;
+                }
+            }
+            // DivineMC end - lithium: sleeping_block_entity
             if (isFullContainer(attachedContainer, opposite)) {
+                // DivineMC start - SparklyPaper: Allow throttling hopper checks if the target container is full
+                if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.hopperThrottleWhenFull && org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.hopperThrottleSkipTicks > 0) {
+                    blockEntity.setCooldown(org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.hopperThrottleSkipTicks);
+                }
+                // DivineMC end - SparklyPaper: Allow throttling hopper checks if the target container is full
                 return false;
             } else {
                 // Paper start - Perf: Optimize Hoppers
@@ -531,10 +_,18 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     public static boolean suckInItems(Level level, Hopper hopper) {
         BlockPos blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
         BlockState blockState = level.getBlockState(blockPos);
-        Container sourceContainer = getSourceContainer(level, hopper, blockPos, blockState);
+        Container sourceContainer = org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity ? getExtractInventory(level, hopper, blockPos, blockState) : getSourceContainer(level, hopper, blockPos, blockState); // DivineMC - lithium: sleeping_block_entity
         if (sourceContainer != null) {
             Direction direction = Direction.DOWN;
-            skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
+            skipPullModeEventFireThreadLocal.set(skipHopperEvents); // Paper - Perf: Optimize Hoppers // ShreddedPaper - use thread local for skipPullModeEventFire
+            // DivineMC start - lithium: sleeping_block_entity
+            if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity) {
+                Boolean res = lithiumExtract(level, hopper, sourceContainer);
+                if (res != null) {
+                    return res;
+                }
+            }
+            // DivineMC end - lithium: sleeping_block_entity
 
             for (int i : getSlots(sourceContainer, direction)) {
                 if (tryTakeInItemFromSlot(hopper, sourceContainer, i, direction, level)) { // Spigot
@@ -546,7 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         } else {
             boolean flag = hopper.isGridAligned() && blockState.isCollisionShapeFullBlock(level, blockPos) && !blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS);
             if (!flag) {
-                for (ItemEntity itemEntity : getItemsAtAndAbove(level, hopper)) {
+                for (ItemEntity itemEntity : org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity ? lithiumGetInputItemEntities(level, hopper) : getItemsAtAndAbove(level, hopper)) { // DivineMC - lithium: sleeping_block_entity
                     if (addItem(hopper, itemEntity)) {
                         return true;
                     }
@@ -683,9 +_,9 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                     stack = stack.split(destination.getMaxStackSize());
                 }
                 // Spigot end
-                ignoreBlockEntityUpdates = true; // Paper - Perf: Optimize Hoppers
+                ignoreBlockEntityUpdatesThreadLocal.set(true); // Paper - Perf: Optimize Hoppers // ShreddedPaper - use thread local for ignoreBlockEntityUpdates
                 destination.setItem(slot, stack);
-                ignoreBlockEntityUpdates = false; // Paper - Perf: Optimize Hoppers
+                ignoreBlockEntityUpdatesThreadLocal.remove(); // Paper - Perf: Optimize Hoppers // ShreddedPaper - use thread local for ignoreBlockEntityUpdates
                 stack = leftover; // Paper - Make hoppers respect inventory max stack size
                 flag = true;
             } else if (canMergeItems(item, stack)) {
@@ -715,7 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
     // CraftBukkit start
     private static @Nullable Container runHopperInventorySearchEvent(
-        Container container,
+        @Nullable Container container,
         org.bukkit.craftbukkit.block.CraftBlock hopper,
         org.bukkit.craftbukkit.block.CraftBlock searchLocation,
         org.bukkit.event.inventory.HopperInventorySearchEvent.ContainerType containerType
@@ -836,6 +_,19 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     }
 
     public void setCooldown(int cooldownTime) {
+        // DivineMC start - lithium: sleeping_block_entity
+        if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity) {
+            if (cooldownTime == 7) {
+                if (this.tickedGameTime == Long.MAX_VALUE) {
+                    this.sleepOnlyCurrentTick();
+                } else {
+                    this.wakeUpNow();
+                }
+            } else if (cooldownTime > 0 && this.sleepingTicker != null) {
+                this.wakeUpNow();
+            }
+        }
+        // DivineMC end - lithium: sleeping_block_entity
         this.cooldownTime = cooldownTime;
     }
 
@@ -855,6 +_,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     @Override
     protected void setItems(NonNullList<ItemStack> items) {
         this.items = items;
+        if (org.bxteam.divinemc.config.DivineConfig.PerformanceCategory.sleepingBlockEntity)  this.lithium$emitStackListReplaced(); // DivineMC - lithium: sleeping_block_entity
     }
 
     public static void entityInside(Level level, BlockPos pos, BlockState state, Entity entity, HopperBlockEntity blockEntity) {
@@ -869,4 +_,750 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     protected AbstractContainerMenu createMenu(int id, Inventory player) {
         return new HopperMenu(id, player, this);
     }
+
+    // DivineMC start - lithium: sleeping_block_entity
+    @Nullable private RebindableTickingBlockEntityWrapper tickWrapper = null;
+    @Nullable private TickingBlockEntity sleepingTicker = null;
+    private long myModCountAtLastInsert, myModCountAtLastExtract, myModCountAtLastItemCollect;
+    private boolean skipNextSleepCheckAfterCooldown = false;
+
+    private net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory insertionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
+    private net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory extractionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
+
+    @Nullable
+    private Container insertBlockInventory, extractBlockInventory;
+
+    @Nullable
+    private LithiumInventory insertInventory, extractInventory;
+    @Nullable
+    private LithiumStackList insertStackList, extractStackList;
+    private long insertStackListModCount, extractStackListModCount;
+
+    @Nullable
+    private List<net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionItemEntityMovementTracker> collectItemEntityTracker;
+    private boolean collectItemEntityTrackerWasEmpty;
+    @Nullable
+    private AABB collectItemEntityBox;
+    private long collectItemEntityAttemptTime;
+
+    @Nullable
+    private List<net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker> extractInventoryEntityTracker;
+    @Nullable
+    private AABB extractInventoryEntityBox;
+    private long extractInventoryEntityFailedSearchTime;
+
+    @Nullable
+    private List<net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker> insertInventoryEntityTracker;
+    @Nullable
+    private AABB insertInventoryEntityBox;
+    private long insertInventoryEntityFailedSearchTime;
+
+    private boolean shouldCheckSleep;
+
+    private void checkSleepingConditions() {
+        if (this.cooldownTime > 0 || this.getLevel() == null || skipNextSleepCheckAfterCooldown) {
+            return;
+        }
+        if (isSleeping()) {
+            return;
+        }
+        if (!this.shouldCheckSleep) {
+            this.shouldCheckSleep = true;
+            return;
+        }
+        boolean listenToExtractTracker = false;
+        boolean listenToInsertTracker = false;
+        boolean listenToExtractEntities = false;
+        boolean listenToItemEntities = false;
+        boolean listenToInsertEntities = false;
+
+        net.caffeinemc.mods.lithium.common.hopper.LithiumStackList thisStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(this);
+
+        if (this.extractionMode != net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE && thisStackList.getFullSlots() != thisStackList.size()) {
+            if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
+                Container blockInventory = this.extractBlockInventory;
+                if (this.extractStackList != null &&
+                    blockInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) {
+                    if (!this.extractStackList.maybeSendsComparatorUpdatesOnFailedExtract() || (blockInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_comparator_tracking.ComparatorTracker comparatorTracker && !comparatorTracker.lithium$hasAnyComparatorNearby())) {
+                        listenToExtractTracker = true;
+                    } else {
+                        return;
+                    }
+                } else {
+                    return;
+                }
+            } else if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY) {
+                BlockState hopperState = this.getBlockState();
+                listenToExtractEntities = true;
+
+                BlockPos blockPos = this.getBlockPos().above();
+                BlockState blockState = this.getLevel().getBlockState(blockPos);
+                if (!blockState.isCollisionShapeFullBlock(this.getLevel(), blockPos) || blockState.is(BlockTags.DOES_NOT_BLOCK_HOPPERS)) {
+                    listenToItemEntities = true;
+                }
+            } else {
+                return;
+            }
+        }
+        if (this.insertionMode != net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE && 0 < thisStackList.getOccupiedSlots()) {
+            if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
+                Container blockInventory = this.insertBlockInventory;
+                if (this.insertStackList != null && blockInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) {
+                    listenToInsertTracker = true;
+                } else {
+                    return;
+                }
+            } else if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY) {
+                BlockState hopperState = this.getBlockState();
+                listenToInsertEntities = true;
+            } else {
+                return;
+            }
+        }
+
+        if (listenToExtractTracker) {
+            ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this.extractBlockInventory).listenForContentChangesOnce(this.extractStackList, this);
+        }
+        if (listenToInsertTracker) {
+            ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this.insertBlockInventory).listenForContentChangesOnce(this.insertStackList, this);
+        }
+        if (listenToInsertEntities) {
+            if (this.insertInventoryEntityTracker == null || this.insertInventoryEntityTracker.isEmpty()) {
+                return;
+            }
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.listenToEntityMovementOnce(this, insertInventoryEntityTracker);
+        }
+        if (listenToExtractEntities) {
+            if (this.extractInventoryEntityTracker == null || this.extractInventoryEntityTracker.isEmpty()) {
+                return;
+            }
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.listenToEntityMovementOnce(this, extractInventoryEntityTracker);
+        }
+        if (listenToItemEntities) {
+            if (this.collectItemEntityTracker == null || this.collectItemEntityTracker.isEmpty()) {
+                return;
+            }
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.listenToEntityMovementOnce(this, collectItemEntityTracker);
+        }
+
+        this.listenForContentChangesOnce(thisStackList, this);
+        lithium$startSleeping();
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(@Nullable TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+
+    @Override
+    public @Nullable TickingBlockEntity lithium$getSleepingTicker() {
+        return sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+        this.lithium$setSleepingTicker(null);
+    }
+
+    @Override
+    public @Nullable RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return tickWrapper;
+    }
+
+    @Override
+    public boolean lithium$startSleeping() {
+        if (this.isSleeping()) {
+            return false;
+        }
+
+        net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper = this.lithium$getTickWrapper();
+        if (tickWrapper != null) {
+            this.lithium$setSleepingTicker(tickWrapper.ticker);
+            tickWrapper.rebind(net.caffeinemc.mods.lithium.common.block.entity.SleepingBlockEntity.SLEEPING_BLOCK_ENTITY_TICKER);
+
+            // Set the last tick time to max value, so other hoppers transferring into this hopper will set it to 7gt
+            // cooldown. Then when waking up, we make sure to not tick this hopper in the same gametick.
+            // This makes the observable hopper cooldown not be different from vanilla.
+            this.tickedGameTime = Long.MAX_VALUE;
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void handleEntityMovement() {
+        this.wakeUpNow();
+    }
+
+    @Override
+    public NonNullList<ItemStack> getInventoryLithium() {
+        return items;
+    }
+
+    @Override
+    public void setInventoryLithium(NonNullList<ItemStack> inventory) {
+        this.items = inventory;
+    }
+
+    @Override
+    public void lithium$handleInventoryContentModified(Container inventory) {
+        wakeUpNow();
+    }
+
+    @Override
+    public void lithium$handleInventoryRemoved(Container inventory) {
+        wakeUpNow();
+        if (inventory == this.insertBlockInventory) {
+            this.invalidateBlockInsertionData();
+        }
+        if (inventory == this.extractBlockInventory) {
+            this.invalidateBlockExtractionData();
+        }
+        if (inventory == this) {
+            this.invalidateCachedData();
+        }
+    }
+
+    @Override
+    public boolean lithium$handleComparatorAdded(Container inventory) {
+        if (inventory == this.extractBlockInventory) {
+            wakeUpNow();
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void lithium$invalidateCacheOnNeighborUpdate(boolean fromAbove) {
+        if (fromAbove) {
+            if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY || this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE) {
+                this.invalidateBlockExtractionData();
+            }
+        } else {
+            if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY || this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE) {
+                this.invalidateBlockInsertionData();
+            }
+        }
+    }
+
+    @Override
+    public void lithium$invalidateCacheOnUndirectedNeighborUpdate() {
+        if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY || this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE) {
+            this.invalidateBlockExtractionData();
+        }
+        if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY || this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE) {
+            this.invalidateBlockInsertionData();
+        }
+    }
+
+    @Override
+    public void lithium$invalidateCacheOnNeighborUpdate(Direction fromDirection) {
+        boolean fromAbove = fromDirection == Direction.UP;
+        if (fromAbove || this.getBlockState().getValue(HopperBlock.FACING) == fromDirection) {
+            this.lithium$invalidateCacheOnNeighborUpdate(fromAbove);
+        }
+    }
+
+    private void invalidateBlockInsertionData() {
+        this.insertionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
+        this.insertBlockInventory = null;
+        this.insertInventory = null;
+        this.insertStackList = null;
+        this.insertStackListModCount = 0;
+
+        wakeUpNow();
+    }
+
+    private void invalidateCachedData() {
+        this.shouldCheckSleep = false;
+        this.invalidateInsertionData();
+        this.invalidateExtractionData();
+    }
+
+    private void invalidateInsertionData() {
+        if (this.level instanceof net.minecraft.server.level.ServerLevel) {
+            if (this.insertInventoryEntityTracker != null) {
+                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.insertInventoryEntityTracker);
+                this.insertInventoryEntityTracker = null;
+                this.insertInventoryEntityBox = null;
+                this.insertInventoryEntityFailedSearchTime = 0L;
+            }
+        }
+
+        if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
+            assert this.insertBlockInventory != null;
+            ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this.insertBlockInventory).stopListenForMajorInventoryChanges(this);
+        }
+        this.invalidateBlockInsertionData();
+    }
+
+    private void invalidateExtractionData() {
+        if (this.level instanceof net.minecraft.server.level.ServerLevel) {
+            if (this.extractInventoryEntityTracker != null) {
+                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.extractInventoryEntityTracker);
+                this.extractInventoryEntityTracker = null;
+                this.extractInventoryEntityBox = null;
+                this.extractInventoryEntityFailedSearchTime = 0L;
+            }
+            if (this.collectItemEntityTracker != null) {
+                net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.unregister(this.collectItemEntityTracker);
+                this.collectItemEntityTracker = null;
+                this.collectItemEntityBox = null;
+                this.collectItemEntityTrackerWasEmpty = false;
+            }
+        }
+        if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
+            assert this.extractBlockInventory != null;
+            ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) this.extractBlockInventory).stopListenForMajorInventoryChanges(this);
+        }
+        this.invalidateBlockExtractionData();
+    }
+
+    private void invalidateBlockExtractionData() {
+        this.extractionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
+        this.extractBlockInventory = null;
+        this.extractInventory = null;
+        this.extractStackList = null;
+        this.extractStackListModCount = 0;
+
+        this.wakeUpNow();
+    }
+
+    private static @Nullable Container getExtractInventory(Level world, Hopper hopper, BlockPos extractBlockPos, BlockState extractBlockState) {
+        if (!(hopper instanceof HopperBlockEntity hopperBlockEntity)) {
+            return getSourceContainer(world, hopper, extractBlockPos, extractBlockState); //Hopper Minecarts do not cache Inventories
+        }
+
+        Container blockInventory = hopperBlockEntity.lithium$getExtractBlockInventory(world, extractBlockPos, extractBlockState);
+        if (blockInventory == null) {
+            blockInventory = hopperBlockEntity.lithium$getExtractEntityInventory(world);
+        }
+        return org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0 ? blockInventory : runHopperInventorySearchEvent(
+            blockInventory,
+            org.bukkit.craftbukkit.block.CraftBlock.at(world, hopperBlockEntity.getBlockPos()),
+            org.bukkit.craftbukkit.block.CraftBlock.at(world, extractBlockPos),
+            org.bukkit.event.inventory.HopperInventorySearchEvent.ContainerType.SOURCE
+        );
+    }
+
+    public @Nullable Container lithium$getExtractBlockInventory(Level world, BlockPos extractBlockPos, BlockState extractBlockState) {
+        Container blockInventory = this.extractBlockInventory;
+        if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY) {
+            return null;
+        } else if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE) {
+            return blockInventory;
+        } else if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
+            return blockInventory;
+        } else if (this.extractionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_ENTITY) {
+            BlockEntity blockEntity = (BlockEntity) java.util.Objects.requireNonNull(blockInventory);
+            BlockPos pos = blockEntity.getBlockPos();
+            if (!(blockEntity).isRemoved() && pos.equals(extractBlockPos)) {
+                net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory;
+                if ((optimizedInventory = this.extractInventory) != null) {
+                    net.caffeinemc.mods.lithium.common.hopper.LithiumStackList insertInventoryStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(optimizedInventory);
+                    if (insertInventoryStackList == this.extractStackList) {
+                        return optimizedInventory;
+                    } else {
+                        this.invalidateBlockExtractionData();
+                    }
+                } else {
+                    return blockInventory;
+                }
+            }
+        }
+
+        blockInventory = getBlockContainer(world, extractBlockPos, extractBlockState);
+        blockInventory = net.caffeinemc.mods.lithium.common.hopper.HopperHelper.replaceDoubleInventory(blockInventory);
+        this.cacheExtractBlockInventory(blockInventory);
+        return blockInventory;
+    }
+
+    public @Nullable Container lithium$getInsertBlockInventory(Level world) {
+        Container blockInventory = this.insertBlockInventory;
+        if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY) {
+            return null;
+        } else if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE) {
+            return blockInventory;
+        } else if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY) {
+            return blockInventory;
+        } else if (this.insertionMode == net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_ENTITY) {
+            BlockEntity blockEntity = (BlockEntity) java.util.Objects.requireNonNull(blockInventory);
+            //Movable Block Entity compatibility - position comparison
+            BlockPos pos = blockEntity.getBlockPos();
+            Direction direction = this.facing;
+            BlockPos transferPos = this.getBlockPos().relative(direction);
+            if (!(blockEntity).isRemoved() &&
+                pos.equals(transferPos)) {
+                net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory;
+                if ((optimizedInventory = this.insertInventory) != null) {
+                    net.caffeinemc.mods.lithium.common.hopper.LithiumStackList insertInventoryStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(optimizedInventory);
+                    //This check is necessary as sometimes the stacklist is silently replaced (e.g. command making furnace read inventory from nbt)
+                    if (insertInventoryStackList == this.insertStackList) {
+                        return optimizedInventory;
+                    } else {
+                        this.invalidateBlockInsertionData();
+                    }
+                } else {
+                    return blockInventory;
+                }
+            }
+        }
+
+        //No Cached Inventory: Get like vanilla and cache
+        Direction direction = this.facing;
+        BlockPos insertBlockPos = this.getBlockPos().relative(direction);
+        BlockState blockState = world.getBlockState(insertBlockPos);
+        blockInventory = getBlockContainer(world, insertBlockPos, blockState);
+        blockInventory = net.caffeinemc.mods.lithium.common.hopper.HopperHelper.replaceDoubleInventory(blockInventory);
+        this.cacheInsertBlockInventory(blockInventory);
+        return blockInventory;
+    }
+
+    public @Nullable Container getInsertInventory(Level world) {
+        Container blockInventory = getInsertInventory0(world);
+        return org.bukkit.event.inventory.HopperInventorySearchEvent.getHandlerList().getRegisteredListeners().length == 0 ? blockInventory : runHopperInventorySearchEvent(
+            blockInventory,
+            org.bukkit.craftbukkit.block.CraftBlock.at(world, this.getBlockPos()),
+            org.bukkit.craftbukkit.block.CraftBlock.at(world, this.getBlockPos().relative(this.facing)),
+            org.bukkit.event.inventory.HopperInventorySearchEvent.ContainerType.DESTINATION
+        );
+    }
+
+    public @Nullable Container getInsertInventory0(Level world) {
+        Container blockInventory = this.lithium$getInsertBlockInventory(world);
+        if (blockInventory != null) {
+            return blockInventory;
+        }
+
+        if (this.insertInventoryEntityTracker == null) {
+            this.initInsertInventoryTracker(world);
+        }
+        if (net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(this.insertInventoryEntityFailedSearchTime, this.insertInventoryEntityTracker)) {
+            this.insertInventoryEntityFailedSearchTime = this.tickedGameTime;
+            return null;
+        }
+        this.insertInventoryEntityFailedSearchTime = Long.MIN_VALUE;
+        this.shouldCheckSleep = false;
+
+        List<Container> inventoryEntities = net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker.getEntities(world, this.insertInventoryEntityBox);
+        if (inventoryEntities.isEmpty()) {
+            this.insertInventoryEntityFailedSearchTime = this.tickedGameTime;
+            //Remember failed entity search timestamp. This allows shortcutting if no entity movement happens.
+            return null;
+        }
+        Container inventory = inventoryEntities.get(world.random.nextInt(inventoryEntities.size()));
+        if (inventory instanceof net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory) {
+            net.caffeinemc.mods.lithium.common.hopper.LithiumStackList insertInventoryStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(optimizedInventory);
+            if (inventory != this.insertInventory || this.insertStackList != insertInventoryStackList) {
+                this.cacheInsertLithiumInventory(optimizedInventory);
+            }
+        }
+
+        return inventory;
+    }
+
+    private void initCollectItemEntityTracker() {
+        assert this.level instanceof net.minecraft.server.level.ServerLevel;
+        AABB inputBox = this.getSuckAabb().move(this.worldPosition.getX(), this.worldPosition.getY(), this.worldPosition.getZ());
+        this.collectItemEntityBox = inputBox;
+        this.collectItemEntityTracker =
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionItemEntityMovementTracker.registerAt(
+                (net.minecraft.server.level.ServerLevel) this.level,
+                inputBox
+            );
+        this.collectItemEntityAttemptTime = Long.MIN_VALUE;
+    }
+
+    private void initExtractInventoryTracker(Level world) {
+        assert world instanceof net.minecraft.server.level.ServerLevel;
+        BlockPos pos = this.worldPosition.relative(Direction.UP);
+        this.extractInventoryEntityBox = new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1);
+        this.extractInventoryEntityTracker =
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker.registerAt(
+                (net.minecraft.server.level.ServerLevel) this.level,
+                this.extractInventoryEntityBox
+            );
+        this.extractInventoryEntityFailedSearchTime = Long.MIN_VALUE;
+    }
+
+    private void initInsertInventoryTracker(Level world) {
+        assert world instanceof net.minecraft.server.level.ServerLevel;
+        Direction direction = this.facing;
+        BlockPos pos = this.worldPosition.relative(direction);
+        this.insertInventoryEntityBox = new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1);
+        this.insertInventoryEntityTracker =
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker.registerAt(
+                (net.minecraft.server.level.ServerLevel) this.level,
+                this.insertInventoryEntityBox
+            );
+        this.insertInventoryEntityFailedSearchTime = Long.MIN_VALUE;
+    }
+
+    private @Nullable Container lithium$getExtractEntityInventory(Level world) {
+        if (this.extractInventoryEntityTracker == null) {
+            this.initExtractInventoryTracker(world);
+        }
+        if (net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(this.extractInventoryEntityFailedSearchTime, this.extractInventoryEntityTracker)) {
+            this.extractInventoryEntityFailedSearchTime = this.tickedGameTime;
+            return null;
+        }
+        this.extractInventoryEntityFailedSearchTime = Long.MIN_VALUE;
+        this.shouldCheckSleep = false;
+
+        List<Container> inventoryEntities = net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionInventoryEntityTracker.getEntities(world, this.extractInventoryEntityBox);
+        if (inventoryEntities.isEmpty()) {
+            this.extractInventoryEntityFailedSearchTime = this.tickedGameTime;
+            //only set unchanged when no entity present. this allows shortcutting this case
+            //shortcutting the entity present case requires checking its change counter
+            return null;
+        }
+        Container inventory = inventoryEntities.get(world.random.nextInt(inventoryEntities.size()));
+        if (inventory instanceof net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory) {
+            net.caffeinemc.mods.lithium.common.hopper.LithiumStackList extractInventoryStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(optimizedInventory);
+            if (inventory != this.extractInventory || this.extractStackList != extractInventoryStackList) {
+                //not caching the inventory (NO_BLOCK_INVENTORY prevents it)
+                //make change counting on the entity inventory possible, without caching it as block inventory
+                this.cacheExtractLithiumInventory(optimizedInventory);
+            }
+        }
+        return inventory;
+    }
+
+    /**
+     * Makes this hopper remember the given inventory.
+     *
+     * @param insertInventory Block inventory / Blockentity inventory to be remembered
+     */
+    private void cacheInsertBlockInventory(@Nullable Container insertInventory) {
+        assert !(insertInventory instanceof Entity);
+        if (insertInventory instanceof net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory) {
+            this.cacheInsertLithiumInventory(optimizedInventory);
+        } else {
+            this.insertInventory = null;
+            this.insertStackList = null;
+            this.insertStackListModCount = 0;
+        }
+
+        if (insertInventory instanceof BlockEntity || insertInventory instanceof net.minecraft.world.CompoundContainer) {
+            this.insertBlockInventory = insertInventory;
+            if (insertInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) {
+                this.insertionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY;
+                ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) insertInventory).listenForMajorInventoryChanges(this);
+            } else {
+                this.insertionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_ENTITY;
+            }
+        } else {
+            if (insertInventory == null) {
+                this.insertBlockInventory = null;
+                this.insertionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY;
+            } else {
+                this.insertBlockInventory = insertInventory;
+                this.insertionMode = insertInventory instanceof net.caffeinemc.mods.lithium.common.hopper.BlockStateOnlyInventory ? net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE : net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
+            }
+        }
+    }
+
+    private void cacheInsertLithiumInventory(net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory) {
+        net.caffeinemc.mods.lithium.common.hopper.LithiumStackList insertInventoryStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(optimizedInventory);
+        this.insertInventory = optimizedInventory;
+        this.insertStackList = insertInventoryStackList;
+        this.insertStackListModCount = insertInventoryStackList.getModCount() - 1;
+    }
+
+    private void cacheExtractLithiumInventory(net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory) {
+        net.caffeinemc.mods.lithium.common.hopper.LithiumStackList extractInventoryStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(optimizedInventory);
+        this.extractInventory = optimizedInventory;
+        this.extractStackList = extractInventoryStackList;
+        this.extractStackListModCount = extractInventoryStackList.getModCount() - 1;
+    }
+
+    /**
+     * Makes this hopper remember the given inventory.
+     *
+     * @param extractInventory Block inventory / Blockentity inventory to be remembered
+     */
+    private void cacheExtractBlockInventory(@Nullable Container extractInventory) {
+        assert !(extractInventory instanceof Entity);
+        if (extractInventory instanceof net.caffeinemc.mods.lithium.api.inventory.LithiumInventory optimizedInventory) {
+            this.cacheExtractLithiumInventory(optimizedInventory);
+        } else {
+            this.extractInventory = null;
+            this.extractStackList = null;
+            this.extractStackListModCount = 0;
+        }
+
+        if (extractInventory instanceof BlockEntity || extractInventory instanceof net.minecraft.world.CompoundContainer) {
+            this.extractBlockInventory = extractInventory;
+            if (extractInventory instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) {
+                this.extractionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.REMOVAL_TRACKING_BLOCK_ENTITY;
+                ((net.caffeinemc.mods.lithium.common.block.entity.inventory_change_tracking.InventoryChangeTracker) extractInventory).listenForMajorInventoryChanges(this);
+            } else {
+                this.extractionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_ENTITY;
+            }
+        } else {
+            if (extractInventory == null) {
+                this.extractBlockInventory = null;
+                this.extractionMode = net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.NO_BLOCK_INVENTORY;
+            } else {
+                this.extractBlockInventory = extractInventory;
+                this.extractionMode = extractInventory instanceof net.caffeinemc.mods.lithium.common.hopper.BlockStateOnlyInventory ? net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.BLOCK_STATE : net.caffeinemc.mods.lithium.common.hopper.HopperCachingState.BlockInventory.UNKNOWN;
+            }
+        }
+    }
+
+    private static List<ItemEntity> lithiumGetInputItemEntities(Level world, Hopper hopper) {
+        if (!(hopper instanceof HopperBlockEntity hopperBlockEntity)) {
+            return getItemsAtAndAbove(world, hopper); //optimizations not implemented for hopper minecarts
+        }
+
+        if (hopperBlockEntity.collectItemEntityTracker == null) {
+            hopperBlockEntity.initCollectItemEntityTracker();
+        }
+
+        long modCount = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(hopperBlockEntity).getModCount();
+
+        if ((hopperBlockEntity.collectItemEntityTrackerWasEmpty || hopperBlockEntity.myModCountAtLastItemCollect == modCount) &&
+            net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionEntityMovementTracker.isUnchangedSince(hopperBlockEntity.collectItemEntityAttemptTime, hopperBlockEntity.collectItemEntityTracker)) {
+            hopperBlockEntity.collectItemEntityAttemptTime = hopperBlockEntity.tickedGameTime;
+            return java.util.Collections.emptyList();
+        }
+
+        hopperBlockEntity.myModCountAtLastItemCollect = modCount;
+        hopperBlockEntity.shouldCheckSleep = false;
+
+        List<ItemEntity> itemEntities = net.caffeinemc.mods.lithium.common.tracking.entity.ChunkSectionItemEntityMovementTracker.getEntities(world, hopperBlockEntity.collectItemEntityBox);
+        hopperBlockEntity.collectItemEntityAttemptTime = hopperBlockEntity.tickedGameTime;
+        hopperBlockEntity.collectItemEntityTrackerWasEmpty = itemEntities.isEmpty();
+        //set unchanged so that if this extract fails and there is no other change to hoppers or items, extracting
+        // items can be skipped.
+        return itemEntities;
+    }
+
+    private static @Nullable Boolean lithiumInsert(Level world, BlockPos pos, HopperBlockEntity hopperBlockEntity, @Nullable Container insertInventory) {
+        if (insertInventory == null || hopperBlockEntity instanceof net.minecraft.world.WorldlyContainer) {
+            //call the vanilla code to allow other mods inject features
+            //e.g. carpet mod allows hoppers to insert items into wool blocks
+            return null;
+        }
+
+        net.caffeinemc.mods.lithium.common.hopper.LithiumStackList hopperStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(hopperBlockEntity);
+        if (hopperBlockEntity.insertInventory == insertInventory && hopperStackList.getModCount() == hopperBlockEntity.myModCountAtLastInsert) {
+            if (hopperBlockEntity.insertStackList != null && hopperBlockEntity.insertStackList.getModCount() == hopperBlockEntity.insertStackListModCount) {
+                //ComparatorUpdatePattern.NO_UPDATE.apply(hopperBlockEntity, hopperStackList); //commented because it's a noop, Hoppers do not send useless comparator updates
+                return false;
+            }
+        }
+
+        boolean insertInventoryWasEmptyHopperNotDisabled = insertInventory instanceof HopperBlockEntity hopperInv &&
+            !hopperInv.isOnCustomCooldown() && hopperBlockEntity.insertStackList != null &&
+            hopperBlockEntity.insertStackList.getOccupiedSlots() == 0;
+
+        boolean insertInventoryHandlesModdedCooldown =
+            insertInventory.canReceiveTransferCooldown() &&
+                hopperBlockEntity.insertStackList != null ?
+                hopperBlockEntity.insertStackList.getOccupiedSlots() == 0 :
+                insertInventory.isEmpty();
+
+            skipPushModeEventFireThreadLocal.set(skipHopperEvents); // ShreddedPaper - use thread local for skipPushModeEventFire
+        //noinspection ConstantConditions
+        if (!(hopperBlockEntity.insertInventory == insertInventory && hopperBlockEntity.insertStackList.getFullSlots() == hopperBlockEntity.insertStackList.size())) {
+            Direction fromDirection = hopperBlockEntity.facing.getOpposite();
+            int size = hopperStackList.size();
+            for (int i = 0; i < size; ++i) {
+                ItemStack transferStack = hopperStackList.get(i);
+                if (!transferStack.isEmpty()) {
+                if (!skipPushModeEventFireThreadLocal.get() && canTakeItemFromContainer(insertInventory, hopperBlockEntity, transferStack, i, Direction.DOWN)) { // ShreddedPaper - use thread local for skipPushModeEventFire
+                        transferStack = callPushMoveEvent(insertInventory, transferStack, hopperBlockEntity);
+                        if (transferStack == null) { // cancelled
+                            break;
+                        }
+                    }
+                    boolean transferSuccess = net.caffeinemc.mods.lithium.common.hopper.HopperHelper.tryMoveSingleItem(insertInventory, transferStack, fromDirection);
+                    if (transferSuccess) {
+                        if (insertInventoryWasEmptyHopperNotDisabled) {
+                            HopperBlockEntity receivingHopper = (HopperBlockEntity) insertInventory;
+                            int k = 8;
+                            if (receivingHopper.tickedGameTime >= hopperBlockEntity.tickedGameTime) {
+                                k = 7;
+                            }
+                            receivingHopper.setCooldown(k);
+                        }
+                        if (insertInventoryHandlesModdedCooldown) {
+                            insertInventory.setTransferCooldown(hopperBlockEntity.tickedGameTime);
+                        }
+                        insertInventory.setChanged();
+                        return true;
+                    }
+                }
+            }
+        }
+        hopperBlockEntity.myModCountAtLastInsert = hopperStackList.getModCount();
+        if (hopperBlockEntity.insertStackList != null) {
+            hopperBlockEntity.insertStackListModCount = hopperBlockEntity.insertStackList.getModCount();
+        }
+        return false;
+    }
+
+    private static @Nullable Boolean lithiumExtract(Level world, Hopper to, Container from) {
+        if (!(to instanceof HopperBlockEntity hopperBlockEntity)) {
+            return null; //optimizations not implemented for hopper minecarts
+        }
+
+        if (from != hopperBlockEntity.extractInventory || hopperBlockEntity.extractStackList == null) {
+            return null; //from inventory is not an optimized inventory, vanilla fallback
+        }
+
+        net.caffeinemc.mods.lithium.common.hopper.LithiumStackList hopperStackList = net.caffeinemc.mods.lithium.common.hopper.InventoryHelper.getLithiumStackList(hopperBlockEntity);
+        net.caffeinemc.mods.lithium.common.hopper.LithiumStackList fromStackList = hopperBlockEntity.extractStackList;
+
+        if (hopperStackList.getModCount() == hopperBlockEntity.myModCountAtLastExtract) {
+            if (fromStackList.getModCount() == hopperBlockEntity.extractStackListModCount) {
+                if (!(from instanceof net.caffeinemc.mods.lithium.common.block.entity.inventory_comparator_tracking.ComparatorTracker comparatorTracker) || comparatorTracker.lithium$hasAnyComparatorNearby()) {
+                    //noinspection CollectionAddedToSelf
+                    fromStackList.runComparatorUpdatePatternOnFailedExtract(fromStackList, from);
+                }
+                return false;
+            }
+        }
+
+        int[] availableSlots = from instanceof WorldlyContainer ? ((WorldlyContainer) from).getSlotsForFace(Direction.DOWN) : null;
+        int fromSize = availableSlots != null ? availableSlots.length : from.getContainerSize();
+        for (int i = 0; i < fromSize; i++) {
+            int fromSlot = availableSlots != null ? availableSlots[i] : i;
+            ItemStack itemStack = fromStackList.get(fromSlot);
+            if (!itemStack.isEmpty() && canTakeItemFromContainer(to, from, itemStack, fromSlot, Direction.DOWN)) {
+                if (!skipPullModeEventFireThreadLocal.get()) { // ShreddedPaper - use thread local for skipPullModeEventFire
+                    itemStack = callPullMoveEvent(to, from, itemStack);
+                    if (itemStack == null) { // cancelled
+                        return true;
+                    }
+                }
+                //calling removeStack is necessary due to its side effects (markDirty in LootableContainerBlockEntity)
+                ItemStack takenItem = from.removeItem(fromSlot, 1);
+                assert !takenItem.isEmpty();
+                boolean transferSuccess = net.caffeinemc.mods.lithium.common.hopper.HopperHelper.tryMoveSingleItem(to, takenItem, null);
+                if (transferSuccess) {
+                    to.setChanged();
+                    from.setChanged();
+                    return true;
+                }
+                //put the item back similar to vanilla
+                ItemStack restoredStack = fromStackList.get(fromSlot);
+                if (restoredStack.isEmpty()) {
+                    restoredStack = takenItem;
+                } else {
+                    restoredStack.grow(1);
+                }
+                //calling setStack is necessary due to its side effects (markDirty in LootableContainerBlockEntity)
+                from.setItem(fromSlot, restoredStack);
+            }
+        }
+        hopperBlockEntity.myModCountAtLastExtract = hopperStackList.getModCount();
+        if (fromStackList != null) {
+            hopperBlockEntity.extractStackListModCount = fromStackList.getModCount();
+        }
+        return false;
+    }
+    // DivineMC end - lithium: sleeping_block_entity
 }