--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
@@ -1,16 +_,16 @@
 package ca.spottedleaf.moonrise.patches.chunk_system.io;
 
 import net.minecraft.nbt.CompoundTag;
-import net.minecraft.world.level.chunk.storage.RegionFile;
+import org.bxteam.divinemc.region.IRegionFile;
 import java.io.IOException;
 
 public interface ChunkSystemRegionFileStorage {
 
     public boolean moonrise$doesRegionFileNotExistNoIO(final int chunkX, final int chunkZ);
 
-    public RegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ);
+    public IRegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ);
 
-    public RegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException;
+    public IRegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException;
 
     public MoonriseRegionFileIO.RegionDataController.WriteData moonrise$startWrite(
             final int chunkX, final int chunkZ, final CompoundTag compound
