--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -67,8 +_,8 @@
     protected boolean scanForWater(BlockGetter level, BlockPos pos) {
         if (!((net.minecraft.world.level.LevelAccessor) level).getMinecraftWorld().purpurConfig.coralDieOutsideWater) return true; // Purpur - Config to not let coral die
         for (Direction direction : Direction.values()) {
-            FluidState fluidState = level.getFluidState(pos.relative(direction));
-            if (fluidState.is(FluidTags.WATER)) {
+            FluidState fluidState = level.getFluidIfLoaded(pos.relative(direction)); // ShreddedPaper - avoid sync chunk loads from scheduled coral ticks
+            if (fluidState == null || fluidState.is(FluidTags.WATER)) { // ShreddedPaper - avoid sync chunk loads from scheduled coral ticks
                 return true;
             }
         }
