--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -191,12 +_,17 @@
 
     private boolean shouldSpreadLiquid(Level level, BlockPos pos, BlockState state) {
         if (this.fluid.is(FluidTags.LAVA)) {
-            boolean isSoulSoil = level.getBlockState(pos.below()).is(Blocks.SOUL_SOIL);
+            BlockState belowState = level.getBlockStateIfLoaded(pos.below()); // ShreddedPaper - avoid sync chunk loads during independent region post-processing
+            boolean isSoulSoil = belowState != null && belowState.is(Blocks.SOUL_SOIL);
 
             for (Direction direction : POSSIBLE_FLOW_DIRECTIONS) {
                 BlockPos blockPos = pos.relative(direction.getOpposite());
-                if (level.getFluidState(blockPos).is(FluidTags.WATER)) {
-                    Block block = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
+                BlockState neighborState = level.getBlockStateIfLoaded(blockPos);
+                if (neighborState == null) {
+                    continue;
+                }
+                if (neighborState.getFluidState().is(FluidTags.WATER)) {
+                    Block block = state.getFluidState().isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, block.defaultBlockState(), Block.UPDATE_ALL)) {
                         this.fizz(level, pos);
@@ -205,7 +_,7 @@
                     return false;
                 }
 
-                if (isSoulSoil && level.getBlockState(blockPos).is(Blocks.BLUE_ICE)) {
+                if (isSoulSoil && neighborState.is(Blocks.BLUE_ICE)) {
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, Blocks.BASALT.defaultBlockState(), Block.UPDATE_ALL)) {
                         this.fizz(level, pos);
