--- a/src/main/java/io/papermc/paper/connection/HorriblePlayerLoginEventHack.java
+++ b/src/main/java/io/papermc/paper/connection/HorriblePlayerLoginEventHack.java
@@ -4,2 +4,3 @@
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.privacy.AnonymousMode;
 import io.papermc.paper.adventure.PaperAdventure;
@@ -86,3 +87,8 @@
         CraftPlayer horribleBukkitPlayer = player.getBukkitEntity();
-        PlayerLoginEvent event = new PlayerLoginEvent(horribleBukkitPlayer, connection.hostname, ((java.net.InetSocketAddress) connection.getRemoteAddress()).getAddress(), ((java.net.InetSocketAddress) connection.channel.remoteAddress()).getAddress());
+        PlayerLoginEvent event = new PlayerLoginEvent(
+            horribleBukkitPlayer,
+            connection.hostname,
+            AnonymousMode.redactLoginEventAddress(((java.net.InetSocketAddress) connection.getRemoteAddress()).getAddress()),
+            AnonymousMode.redactLoginEventAddress(((java.net.InetSocketAddress) connection.channel.remoteAddress()).getAddress())
+        );
         event.disallow(result.result(), PaperAdventure.asAdventure(result.message()));
