--- a/src/main/java/io/papermc/paper/connection/PaperCommonConnection.java
+++ b/src/main/java/io/papermc/paper/connection/PaperCommonConnection.java
@@ -5,2 +5,3 @@
 import com.google.common.base.Preconditions;
+import io.multipaper.shreddedpaper.privacy.AnonymousMode;
 import io.papermc.paper.adventure.PaperAdventure;
@@ -84,3 +85,3 @@
     public SocketAddress getAddress() {
-        return this.handle.connection.channel.remoteAddress();
+        return AnonymousMode.redactPublicSocketAddress(this.handle.connection.channel.remoteAddress());
     }
@@ -89,3 +90,3 @@
     public InetSocketAddress getClientAddress() {
-        return (InetSocketAddress) this.handle.connection.getRemoteAddress();
+        return AnonymousMode.redactPublicSocketAddress((InetSocketAddress) this.handle.connection.getRemoteAddress());
     }
@@ -99,3 +100,5 @@
     public @Nullable InetSocketAddress getHAProxyAddress() {
-        return this.handle.connection.haProxyAddress instanceof final InetSocketAddress inetSocketAddress ? inetSocketAddress : null;
+        return this.handle.connection.haProxyAddress instanceof final InetSocketAddress inetSocketAddress
+            ? AnonymousMode.redactPublicSocketAddress(inetSocketAddress)
+            : null;
     }
