--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -17,6 +_,6 @@
 import org.jspecify.annotations.Nullable;
 
-public interface Container extends Clearable, SlotProvider, Iterable<ItemStack> {
+public interface Container extends Clearable, SlotProvider, Iterable<ItemStack>, net.caffeinemc.mods.lithium.api.inventory.LithiumCooldownReceivingInventory, net.caffeinemc.mods.lithium.api.inventory.LithiumTransferConditionInventory { // ShreddedPaper - Lithium: inventory bridge
     float DEFAULT_DISTANCE_BUFFER = 4.0F;
 
     int getContainerSize();
