--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -16,7 +_,7 @@ import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import org.jspecify.annotations.Nullable;
 
-public interface Container extends Clearable, SlotProvider, Iterable<ItemStack> {
+public interface Container extends Clearable, SlotProvider, Iterable<ItemStack>, net.caffeinemc.mods.lithium.api.inventory.LithiumCooldownReceivingInventory, net.caffeinemc.mods.lithium.api.inventory.LithiumTransferConditionInventory { // DivineMC - lithium: sleeping_block_entity
     float DEFAULT_DISTANCE_BUFFER = 4.0F;
 
     int getContainerSize();